### PR TITLE
DR-89 implement dataset flights and wire up controller

### PR DESCRIPTION
**Pull request by ddietterich**
This PR includes:
- flights for dataset create and delete
- changes to swagger for dataset API: provide correct return model definition; add offset and limit parameters to enumerate
- generic method added to DaoKeyHolder for code reuse
- add missing catch and add some logging in stairway; that had resulted in the stairway thread exiting on a throw without a trace.

You will see that I changed from Instance to Timestamp for dataset. That is probably not the right end state. We will fix that when we do DR-156.

### CODE REVIEW

File: src/main/java/bio/terra/controller/RepositoryApiController.java [hunk 1]
```diff
@@ -119,26 +123,63 @@ protected void initBinder(final WebDataBinder binder) {
...
+    public ResponseEntity<DatasetModel> retrieveDataset(@PathVariable("id") String id) {
+        DatasetModel datasetModel = datasetService.retrieveDataset(UUID.fromString(id));
+        return new ResponseEntity<>(datasetModel, HttpStatus.OK);
+    }
+
+    // -- jobs --
+    @Override
+    public ResponseEntity<List<JobModel>> enumerateJobs(
+            @Valid @RequestParam(value = "offset", required = false, defaultValue = "0") Integer offset,
+            @Valid @RequestParam(value = "limit", required = false, defaultValue = "10") Integer limit) {
```
Comment (thathert) [inline src/main/java/bio/terra/controller/RepositoryApiController.java:171]: Great. We should add this to the list of annotations to pass through when IntelliJ does code generation for overriding a method.

File: src/main/java/bio/terra/stairway/Flight.java [hunk 2]
```diff
@@ -117,6 +122,9 @@ private FlightStatus fly() {
         } catch (InterruptedException ex) {
             Thread.currentThread().interrupt();
             context().setResult(new StepResult(StepStatus.STEP_RESULT_FAILURE_FATAL, ex));
+        } catch (Exception ex) {
+            logger.error("Unhandled step exception", ex);
```
Comment (thathert) [inline src/main/java/bio/terra/stairway/Flight.java:126]: Good catch

File: src/test/java/bio/terra/controller/DatasetOperationTest.java [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        String baseName = datasetRequest.getName();
+        String datasetName = randomizedName(baseName, infix);
+        datasetRequest.setName(datasetName);
+
+        String jsonRequest = objectMapper.writeValueAsString(datasetRequest);
+
+        MvcResult result = mvc.perform(post("/api/repository/v1/datasets")
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(jsonRequest))
+// TODO: swagger field validation errors do not set content type; they log and return nothing
```
Comment (thathert) [inline src/test/java/bio/terra/controller/DatasetOperationTest.java:131]: I added [DR-157](https://broadinstitute.atlassian.net/jira/software/projects/DR/boards/517/backlog?selectedIssue=DR-157) to track.

**Review by thathert — APPROVED** [review]
This looks good to me. @ddietterich have you had a chance to try out the setting in IntelliJ that will selectively copy over annotations from an interface method that you're overriding? I'd like to see if it addresses the issue you ran into last week.

File: src/main/java/bio/terra/controller/RepositoryApiController.java (hunk 1 above)
Comment (ahaessly) [reply src/main/java/bio/terra/controller/RepositoryApiController.java:171]: is the @RequestParam annotation just for documentation or does it actually do or enforce something?

**Review by ahaessly — APPROVED** [review]
just one question

File: src/main/java/bio/terra/controller/RepositoryApiController.java (hunk 1 above)
Comment (ddietterich) [reply src/main/java/bio/terra/controller/RepositoryApiController.java:171]: The annotation is required. It tells Spring how to find the parameters in the request.
I had originally not included the annotations when I generated the @Override of the method. I got some hard-to-understand errors. Jeremy has fixed up the IntelliJ setup to copy these annotations.

**Pull request merged**